### ISSUE TITLE
UB due to unsafe pinning in actix-codec

### ISSUE THREAD

**Issue by sebzim4500** [actix/actix-net#91]
The following code segfaults, due to incorrect pinning in `Framed`. 

```rust
use futures::task::{noop_waker, Context, Poll};
use actix_codec::{Framed, AsyncRead, BytesCodec, AsyncWrite};
use std::pin::Pin;
use futures::io::Error;
use std::future::Future;
use pin_project::pin_project;

#[pin_project]
struct FakeSocket<F> {
    #[pin]
    inner: F
}

impl<F: Future> AsyncRead for FakeSocket<F> {
    fn poll_read(self: Pin<&mut Self>, cx: &mut Context<'_>, buf: &mut [u8]) -> Poll<Result<usize, Error>> {
        self.project().inner.poll(cx).map(|x| Ok(0))
    }
}

impl<F> AsyncWrite for FakeSocket<F> {
    fn poll_write(self: Pin<&mut Self>, cx: &mut Context<'_>, buf: &[u8]) -> Poll<Result<usize, Error>> {
        unimplemented!()
    }

    fn poll_flush(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Error>> {
        unimplemented!()
    }

    fn poll_shutdown(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Error>> {
        unimplemented!()
    }
}

fn main() {
    let (sender, receiver) = futures::channel::oneshot::channel();
    let mut framed: Result<_, [u8; 32]> = Ok(Framed::new(FakeSocket {
        inner: async {
            let x = Box::new(0);
            let y = &x;
            receiver.await.unwrap();
            let z = **y;
        }
    }, BytesCodec));

    let waker = noop_waker();
    let mut context = Context::from_waker(&waker);

    framed.as_mut().unwrap().next_item(&mut context);
    sender.send(()).unwrap();
    let _ = std::mem::replace(&mut framed, Err([0; 32])).unwrap().next_item(&mut context);
}
```

**Comment by robjtede** [comment]
I think this is fixed in c41b5d8dd4235ccca84d0b687996615c0c64d956. The provided code doesn't compile anymore.

**Comment by Shnatsel** [comment]
`pin-project` is sound, and there's no remaining unsafe in the `Framed` implementation, so this indeed should be fixed. Thanks to @nemo157 for confirming.
